### PR TITLE
Refactor search services

### PR DESCRIPTION
**Pull request by stevenjoezhang**
## PR Checklist
**Please check if your PR fulfills the following requirements:**

- [x] The commit message follows [our guidelines](https://github.com/theme-next/hexo-theme-next/blob/master/.github/CONTRIBUTING.md).
- [x] Tests for the changes was maked (for bug fixes / features).
   - [x] Muse | Mist have been tested.
   - [x] Pisces | Gemini have been tested.
- [ ] [Docs](https://github.com/theme-next/theme-next.org/tree/source/source/docs) in [NexT website](https://theme-next.org/docs/) have been added / updated (for features).

## PR Type
**What kind of change does this PR introduce?**

- [ ] Bugfix.
- [ ] Feature.
- [ ] Code style update (formatting, local variables).
- [x] Refactoring (no functional changes, no api changes).
- [ ] Build related changes.
- [ ] CI related changes.
- [ ] Documentation content changes.
- [ ] Other... Please describe:

## What is the current behavior?

Issue resolved: #831

https://juejin.im/post/5bbdc8d8e51d450e5e0cb269

https://github.com/iissnan/hexo-theme-next/pull/694/files

https://github.com/wzpan/hexo-generator-search

https://github.com/wzpan/hexo-theme-freemind/blob/master/source/js/search.js

## What is the new behavior?

- Screenshots with this changes: N/A
- Link to demo site with this changes: N/A

### How to use?
In NexT `_config.yml`:
```yml
...
```

## Does this PR introduce a breaking change?
- [ ] Yes.
- [x] No.

### CODE REVIEW

File: source/js/local-search.js [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+              var content = data.content.trim().replace(/<[^>]+>/g, '');
+              if (CONFIG.localsearch.unescape) {
+                content = unescapeHtml(content);
+              }
+              var contentInLowerCase = content.toLowerCase();
+              var articleUrl = decodeURIComponent(data.url).replace(/\/{2,}/g, '/');
+              var indexOfTitle = [];
+              var indexOfContent = [];
+              // Only match articles with not empty titles
+              if (title !== '') {
```
Comment (jiangtj) [inline source/js/local-search.js]: 这个判断提前到`var title = data.title.trim();`这个前面，如果是生成的是json并且`source/downloads/code`文件夹下有内容，那么title可能为null，`trim()`会报错

File: layout/_partials/search/index.swig [hunk 2]
```diff
@@ -1,7 +1,5 @@
 {% if theme.algolia_search.enable %}
   {% include 'algolia-search.swig' %}
-{% elif theme.swiftype_key %}
-  {% include 'swiftype.swig' %}
 {% elif theme.local_search.enable %}
   {% include 'localsearch.swig' %}
```
Comment (jiangtj) [inline layout/_partials/search/index.swig]: Move ` layout/_partials/search/*` to `layout/_third-party/search/*`?
I don't think it's necessary to have two layouts

Comment (stevenjoezhang) [reply layout/_partials/search/index.swig]: jQuery is loaded at the end of the body. Local search depends on jQuery. Put the script here will cause an `$` not defined error. Maybe we can open another PR to move jQuery to `<head>`

Comment (jiangtj) [reply layout/_partials/search/index.swig]: > Maybe we can open another PR to move jQuery to `<head>`

Don't do this.

我的意思是将原本在菜单那的搜索弹出框布局移到底部，与脚本合并，弹出框与其所在的位置没有关联，两个文件变一个，更便于维护

I mean to move the search popup layout that was originally in the menu to the bottom, merge with the script, the popup is not associated with its location, and the two files become one, which is easier to maintain.

Comment (stevenjoezhang) [reply layout/_partials/search/index.swig]: > 两个文件变一个，更便于维护

但是将DOM和`<script>`分离更加优美。在`layout`中，`<script>`只应存在于`_scripts`和`_third-party`中，各种DOM存在于`_partials`中。就像评论系统，使用inject.comment向评论区写入DOM，使用inject.bodyend写入脚本。它们应该是分开的。

还是需要开一个新的PR，处理这样的情况。

Comment (jiangtj) [reply layout/_partials/search/index.swig]: image
这是因为`comment`的位置比较尴尬，不能在bodyEnd也不是postBodyEnd 😂

Comment (stevenjoezhang) [reply layout/_partials/search/index.swig]: 另外一个问题是各种share，也是同时存在DOM元素和脚本。真蛋疼..

Comment (jiangtj) [reply layout/_partials/search/index.swig]: share完全不知道怎么搞，在我的基于next的主题里是全删了，反正我也不用，哈哈😄

Comment (stevenjoezhang) [reply layout/_partials/search/index.swig]: 所以这咋办呀

Comment (jiangtj) [reply layout/_partials/search/index.swig]: 不知道，那就暂时不管share咯，以后想到怎么处理了在弄。

这是我的搜索重构 https://github.com/jiangtj/hexo-theme-cake/pull/11
加menu注入点，试了把localsearch提取出去，等GitHub包管理出了，可以这样搞啊，反正localsearch本来就要装个hexo插件:joy:

**Pull request merged**